test/lit/passes/remove-unused-brs_branch-hints-shrink.wast

```diff
@@ -0,0 +1,180 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+;; RUN: wasm-opt %s --remove-unused-brs -all --shrink-level=1 -S -o - \
+;; RUN:   | filecheck %s
+
+(module
+  ;; CHECK:      (import "a" "b" (func $none (type $1)))
+  (import "a" "b" (func $none))
+
+  ;; CHECK:      (func $join-br_ifs (type $0) (param $x i32) (param $y i32)
+  ;; CHECK-NEXT:  (block $out
+  ;; CHECK-NEXT:   (@metadata.code.branch_hint "\01")
+  ;; CHECK-NEXT:   (br_if $out
+  ;; CHECK-NEXT:    (i32.or
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:     (local.get $y)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (nop)
+  ;; CHECK-NEXT:   (call $none)
+  ;; CHECK-NEXT:   (br_if $out
+  ;; CHECK-NEXT:    (local.get $y)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $join-br_ifs (param $x i32) (param $y i32)
+    ;; The br_ifs will be joined into a single one. The hint should propagate,
+    ;; as it matches.
+    (block $out
+      (@metadata.code.branch_hint "\01")
+      (br_if $out
+        (local.get $x)
+      )
+      (@metadata.code.branch_hint "\01")
+      (br_if $out
+        (local.get $y)
+      )
+      ;; Extra code so that the entire testcase does not get optimized out as
+      ;; trivial.
+      (call $none)
+      (br_if $out
+        (local.get $y)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $join-br_ifs-0 (type $0) (param $x i32) (param $y i32)
+  ;; CHECK-NEXT:  (block $out
+  ;; CHECK-NEXT:   (@metadata.code.branch_hint "\00")
+  ;; CHECK-NEXT:   (br_if $out
+  ;; CHECK-NEXT:    (i32.or
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:     (local.get $y)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (nop)
+  ;; CHECK-NEXT:   (call $none)
+  ;; CHECK-NEXT:   (br_if $out
+  ;; CHECK-NEXT:    (local.get $y)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $join-br_ifs-0 (param $x i32) (param $y i32)
+    ;; The hints once more match, but now are 0. We still propagate.
+    (block $out
+      (@metadata.code.branch_hint "\00")
+      (br_if $out
+        (local.get $x)
+      )
+      (@metadata.code.branch_hint "\00")
+      (br_if $out
+        (local.get $y)
+      )
+
+      (call $none)
+      (br_if $out
+        (local.get $y)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $join-br_ifs-no (type $0) (param $x i32) (param $y i32)
+  ;; CHECK-NEXT:  (block $out
+  ;; CHECK-NEXT:   (br_if $out
+  ;; CHECK-NEXT:    (i32.or
+  ;; CHECK-NEXT:     (local.get $x)
+  ;; CHECK-NEXT:     (local.get $y)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (nop)
+  ;; CHECK-NEXT:   (call $none)
+  ;; CHECK-NEXT:   (br_if $out
+  ;; CHECK-NEXT:    (local.get $y)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $join-br_ifs-no (param $x i32) (param $y i32)
+    ;; One is missing a hint, so we clear the hint.
+    (block $out
+      (@metadata.code.branch_hint "\00")
```

Comment:
If this hint were a 1, it would make sense to keep, but I don't think we would currently keep it.

Reply:
Good point, this really should be handled. Done (and other feedback too).